Outer product between vectors not work

<!-- We've collected some common issue solutions in https://docs.taichi-lang.org/docs/install. Make sure you've check them out first. Hopefully they could address your problem. -->

**Describe the bug**
Matrix multiplication between n*1 and 1*n vector will cause compilation error.

**To Reproduce**

```py
import taichi as ti
ti.init()

@ti.kernel
def fun():
    a = ti.Vector([0.0, 0.0])
    print(a @ a.transpose())

fun()
```

**Log/Screenshots**

```
{15:04}~/playground:dev ✗ ➭ p test_ra.py
[Taichi] version 1.1.3, llvm 10.0.0, commit 9a35e049, linux, python 3.8.5
[Taichi] Starting on arch=x64
Traceback (most recent call last):
  File "test_ra.py", line 9, in <module>
    fun()
  File "/home/bx2k/git/taichi/python/taichi/lang/kernel_impl.py", line 917, in wrapped
    raise type(e)('\n' + str(e)) from None
taichi.lang.exception.TaichiCompilationError: 
File "test_ra.py", line 7, in fun:
    print(a @ a.transpose())
          ^^^^^^^^^^^^^^^^^
Traceback (most recent call last):
  File "/home/bx2k/git/taichi/python/taichi/lang/ast/ast_transformer_utils.py", line 25, in __call__
    return method(ctx, node)
  File "/home/bx2k/git/taichi/python/taichi/lang/ast/ast_transformer.py", line 738, in build_BinOp
    node.ptr = op(node.left.ptr, node.right.ptr)
  File "/home/bx2k/git/taichi/python/taichi/lang/ast/ast_transformer.py", line 736, in <lambda>
    ast.MatMult: lambda l, r: l @ r,
  File "/home/bx2k/git/taichi/python/taichi/lang/matrix.py", line 564, in __matmul__
    return other.transpose() @ self
  File "/home/bx2k/git/taichi/python/taichi/lang/matrix.py", line 566, in __matmul__
    assert self.m == other.n, f"Dimension mismatch between shapes ({self.n}, {self.m}), ({other.n}, {other.m})"
AssertionError: Dimension mismatch between shapes (2, 1), (2, 1)

```

**Additional comments**
This bug seems to be related to PR #6014 .

--- a/python/taichi/_funcs.py
+++ b/python/taichi/_funcs.py
@@ -3,7 +3,7 @@ import math
 from taichi.lang import impl, matrix, ops
 from taichi.lang.impl import expr_init, get_runtime, grouped, static
 from taichi.lang.kernel_impl import func, pyfunc
-from taichi.lang.matrix import Matrix, Vector, is_vector
+from taichi.lang.matrix import Matrix, Vector
 from taichi.types import f32, f64
 from taichi.types.annotations import template
 

--- a/python/taichi/_funcs.py
+++ b/python/taichi/_funcs.py
@@ -59,9 +59,6 @@ def _matrix_transpose(mat):
     Returns:
         Transpose of the input matrix.
     """
-    if static(is_vector(mat)):
-        # Convert to row vector
-        return matrix.Matrix([[mat(i) for i in range(mat.n)]])
     return matrix.Matrix([[mat(i, j) for i in range(mat.n)]
                           for j in range(mat.m)],
                          ndim=mat.ndim)

--- a/python/taichi/lang/impl.py
+++ b/python/taichi/lang/impl.py
@@ -64,7 +64,8 @@ def expr_init(rhs):
                 entries = [[rhs(i, j) for j in range(rhs.m)]
                            for i in range(rhs.n)]
             return make_matrix(entries)
-        if isinstance(rhs, Vector) or getattr(rhs, "ndim", None) == 1:
+        if (isinstance(rhs, Vector)
+                or getattr(rhs, "ndim", None) == 1) and rhs.m == 1:
             # _IntermediateMatrix may reach here
             return Vector(rhs.to_list(), ndim=rhs.ndim)
         return Matrix(rhs.to_list(), ndim=rhs.ndim)

--- a/python/taichi/lang/matrix.py
+++ b/python/taichi/lang/matrix.py
@@ -669,7 +669,7 @@ class Matrix(TaichiOperations):
         This is similar to `numpy.ndarray`'s `flatten` and `ravel` methods,
         the difference is that this function always returns a new list.
         """
-        if is_vector(self):
+        if is_vector(self) and self.m == 1:
             return [self(i) for i in range(self.n)]
         return [[self(i, j) for j in range(self.m)] for i in range(self.n)]
 

--- a/python/taichi/lang/matrix.py
+++ b/python/taichi/lang/matrix.py
@@ -117,6 +117,10 @@ def is_vector(x):
     return isinstance(x, Vector) or getattr(x, "ndim", None) == 1
 
 
+def is_col_vector(x):
+    return is_vector(x) and getattr(x, "m", None) == 1
+
+
 class _MatrixBaseImpl:
     def __init__(self, m, n, entries):
         self.m = m
@@ -498,7 +502,7 @@ class Matrix(TaichiOperations):
 
     def _element_wise_binary(self, foo, other):
         other = self._broadcast_copy(other)
-        if is_vector(self):
+        if is_col_vector(self):
             return Vector([foo(self(i), other(i)) for i in range(self.n)],
                           ndim=self.ndim)
         return Matrix([[foo(self(i, j), other(i, j)) for j in range(self.m)]
@@ -507,7 +511,7 @@ class Matrix(TaichiOperations):
 
     def _broadcast_copy(self, other):
         if isinstance(other, (list, tuple)):
-            if is_vector(self):
+            if is_col_vector(self):
                 other = Vector(other, ndim=self.ndim)
             else:
                 other = Matrix(other, ndim=self.ndim)
@@ -558,9 +562,9 @@ class Matrix(TaichiOperations):
 
         """
         assert isinstance(other, Matrix), "rhs of `@` is not a matrix / vector"
-        if is_vector(self) and not is_vector(other):
+        if (is_col_vector(self)) and not is_vector(other):
             # left multiplication
-            assert self.n == other.m, f"Dimension mismatch between shapes ({self.n}, {self.m}), ({other.n}, {other.m})"
+            assert self.n == other.m, f"Dimension mismatch between (left multiplication) shapes ({self.n}, {self.m}), ({other.n}, {other.m})"
             return other.transpose() @ self
         # right multiplication
         assert self.m == other.n, f"Dimension mismatch between shapes ({self.n}, {self.m}), ({other.n}, {other.m})"
@@ -572,7 +576,7 @@ class Matrix(TaichiOperations):
                 for k in range(1, other.n):
                     acc = acc + self(i, k) * other(k, j)
                 entries[i].append(acc)
-        if is_vector(other) and other.m == 1:
+        if is_col_vector(other):
             return Vector(entries)
         return Matrix(entries)
 
@@ -669,7 +673,7 @@ class Matrix(TaichiOperations):
         This is similar to `numpy.ndarray`'s `flatten` and `ravel` methods,
         the difference is that this function always returns a new list.
         """
-        if is_vector(self) and self.m == 1:
+        if is_col_vector(self):
             return [self(i) for i in range(self.n)]
         return [[self(i, j) for j in range(self.m)] for i in range(self.n)]
 
@@ -691,7 +695,7 @@ class Matrix(TaichiOperations):
             >>> B
             [0.0, 1.0, 2.0]
         """
-        if is_vector(self):
+        if is_col_vector(self):
             # when using _IntermediateMatrix, we can only check `self.ndim`
             return Vector(
                 [ops_mod.cast(self(i), dtype) for i in range(self.n)])
@@ -1636,8 +1640,8 @@ class MatrixField(Field):
                         (list, tuple)) and isinstance(val[0], numbers.Number):
             assert self.m == 1
             val = tuple(val)
-        elif is_vector(val) or self.ndim == 1:
-            val = tuple([(val(i), ) for i in range(self.n)])
+        elif is_vector(val):
+            val = tuple([(val(i), ) for i in range(self.n * self.m)])
         elif isinstance(val, Matrix):
             val_tuple = []
             for i in range(val.n):

--- a/tests/python/test_matrix.py
+++ b/tests/python/test_matrix.py
@@ -806,3 +806,14 @@ def test_local_matrix_index_check():
     with pytest.raises(TaichiCompilationError,
                        match=r'Expected 1 indices, but got 2'):
         bar()
+
+
+@test_utils.test()
+def test_vector_vector_t():
+    @ti.kernel
+    def foo() -> ti.types.matrix(2, 2, ti.f32):
+        a = ti.Vector([1.0, 2.0])
+        b = ti.Vector([1.0, 2.0])
+        return a @ b.transpose()
+
+    assert foo() == [[1.0, 2.0], [2.0, 4.0]]

